File: zero-query/src/schemas/receipt-user.ts
/**
 * Receipt user schema for Zero.
 *
 * Represents a user that can be assigned to receipt line items for bill splitting.
 * ReceiptUsers can optionally link to a User via user_id, or exist as anonymous
 * users with just a display_name.
 *
 * Database table: receipt_users
 * - Stores receipt user details and optional link to users table.
 * - Supports soft deletion via deleted_at timestamp.
 *
 * Relationships:
 * - Many-to-one with users (receipt_users.user_id -> users.id, optional)
 * - One-to-many with assignments (receipt_users.id -> assignments.receipt_user_id)
 */

import { number, string, table } from "@rocicorp/zero";

export const receiptUser = table("receipt_users")
  .columns({
    id: string(), // ULID, client-generated
    user_id: number().optional(),
    display_name: string().optional(),
    created_at: number(),
    deleted_at: number().optional(),
  })
  .primaryKey("id");

File: backend/models/receipt_user.py
from sqlalchemy import text

from models import db


class ReceiptUser(db.Model):
    __tablename__ = "receipt_users"

    # ID is client-generated ULID via Zero mutators, not auto-incremented
    id = db.Column(db.Text, primary_key=True)
    user_id = db.Column(db.BigInteger, db.ForeignKey("users.id"), nullable=True)
    display_name = db.Column(db.Text, nullable=True)
    created_at = db.Column(
        db.TIMESTAMP(timezone=True), server_default=text("CURRENT_TIMESTAMP")
    )
    deleted_at = db.Column(db.TIMESTAMP(timezone=True), nullable=True, index=True)

    user = db.relationship("User", backref=db.backref("receipt_users", lazy=True))

    def __repr__(self):
        return f"<ReceiptUser {self.id}>"
